### PR TITLE
SSE: Make sure to forward headers, user and cookies/OAuth token

### LINKED ISSUE #513: Math expressions broken with OBO auth (no User and req.Headers)
**What happened**:

When using math expressions with OBO authentication, the `QueryData(ctx, req)` call doesn't receive `req.Headers`.

**What you expected to happen**:

OBO authentication should work and QueryData should receive user information and access token header regardless whether the call originated from a math expression or not.

**How to reproduce it (as minimally and precisely as possible)**:

1. Have #503 patched
2. Open a panel editor with ADX datasource
3. Add any two queries A and B which successfully return data
4. Add an expression, "Math", "$A + $B"
5. Set breakpoint [here](https://github.com/grafana/azure-data-explorer-datasource/blob/e25f75a831651e5825bef183496d778c87624f50/pkg/azuredx/datasource.go#L82)
6. Run math expression, see that value of `req.UserContext.User` is `nil` and `req.Headers` is `nil` too.

**Screenshots**

Error message of a panel containing a math expression:
image

QueryData [here](https://github.com/grafana/azure-data-explorer-datasource/blob/e25f75a831651e5825bef183496d778c87624f50/pkg/azuredx/datasource.go#L82):
image

**Anything else we need to know?**:

This bug is hidden behind the #503 since any expressions are currently broken. But this is another issue which affects OBO authentication.

Related: https://github.com/grafana/grafana/issues/58793

**Environment**:

- Grafana version: main
- Plugin version: main
- OS Grafana is installed on:
- User OS & Browser:
- Others:

### LINKED ISSUE #58793: Math expressions do not pass User info in req.PluginContext
**What happened**:

When using math expressions, the `QueryData(ctx, req)` call doesn't receive user information as part of `req.PluginContext`.

Example in Prometheus datasource ([here](https://github.com/grafana/grafana/blob/93c1fbbe3f3996e2e0d39d39acb235c802d37ed4/pkg/tsdb/prometheus/prometheus.go#L91)) but reproducible in any datasource:

image

Call stack:

image

**What you expected to happen**:

The `User` structure should be filled with user information regardless whether the call originated from math expression or not.

**How to reproduce it (as minimally and precisely as possible)**:

1. Open a panel editor with Prometheus datasource
2. Add any two queries A and B which successfully return data
3. Add an expression, "Math", "$A + $B"
4. Set breakpoint [here](https://github.com/grafana/grafana/blob/93c1fbbe3f3996e2e0d39d39acb235c802d37ed4/pkg/tsdb/prometheus/prometheus.go#L91)
5. Run math expression, see that value of `req.UserContext.User` is `nil`.

**Anything else we need to know?**:

**Environment**:
- Grafana version: main
- Data source type & version: QueryData in any datasource
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by kylebrandt**
Copies the User and Headers from the request through expressions.

possibly fixes #58793 and fixes https://github.com/grafana/azure-data-explorer-datasource/issues/513

**Special notes for your reviewer**:
only first header value per header key is kept

Need someone with passthrough auth setup to verify this works for me.

### CODE REVIEW

**Comment by kylebrandt** [comment]
@kostrse Any chance you can try this branch and see if solves the issues you have with passthrough auth?

**Comment by kostrse** [comment]
@kylebrandt Thanks, I'm doing testing and will reply back once I'm done.

**Review by marefr — APPROVED** [review]
LGTM, added a suggestion

Seems a bit related to https://github.com/grafana/grafana/pull/58132 - need to make sure SSE is considered there.

**Comment by taleena** [comment]
@marefr @papagian looks like this is good to go, but is blocked? Would it also be possible to port this to 9.1, 9.2?

**Comment by kostrse** [comment]
@kylebrandt to repro in ADX

You will need to run Grafana with AAD OAuth configuration, logged in as an AAD user.

You don't need valid credentials for ADX to repro the missing `X-ID-Token` since the code fails before the actual authentication request.

1. Save an ADX datasource with OBO checked.
2. Attach with debugger to ADX backend.
3. Set breakpoint to line 108 of `auth.go` ([here](https://github.com/grafana/azure-data-explorer-datasource/blob/ba4ca9cf75374cd9468f0129c2584dd32cc8b3a5/pkg/azuredx/azureauth/auth.go#L108))
4. Create a new dashboard a a new panel
5. Run query "A"
6. You will hit breakpoint in `QueryData` func and the `req.Headers` will have `X-ID-Token`.
7. Continue
8. Add math exression "B" (e.g. A + 1)
9. You will hit the same breakpoint again but this time `X-ID-Token` will be missing.

The difference is whether math expression is present or not.

**Comment by kostrse** [comment]
cc @andresmgot

File: pkg/services/query/query.go [hunk 1]
```diff
@@ -133,10 +133,31 @@ func (s *Service) QueryData(ctx context.Context, user *user.SignedInUser, skipCa
...
+		clippedHeaders := make(map[string]string, len(parsedReq.httpRequest.Header))
+		for k, v := range parsedReq.httpRequest.Header {
+			if len(v) == 0 {
+				clippedHeaders[k] = ""
+				continue
+			}
+			clippedHeaders[k] = v[0]
+		}
+		exprReq.Headers = clippedHeaders
+	}
```
Comment (marefr) [inline pkg/services/query/query.go:159]: This code feels wrong to me. We don't forward all HTTP headers to `QueryData` today so why do we need to do it for SSE?

**Review by marefr — DISMISSED** [review]
Had a look again. Think there's some wrong assumptions being made in regards to incoming HTTP headers.

**Comment by marefr** [comment]
> @marefr @papagian looks like this is good to go, but is blocked? Would it also be possible to port this to 9.1, 9.2 (with a slight preference to 9.2)?

9.2 sure, but not 9.1 since only security fixes would go into it. Would ask @kylebrandt since he's the author of this PR.

**Comment by marefr** [comment]
To support X-ID-Token in SSE we should reuse the following code for SSE: 
https://github.com/grafana/grafana/blob/c9286ff79464de3e368b3ca95b215a6d674d8437/pkg/services/query/query.go#L201-L231

I would suggest to add this to this PR and get it merged and backported for 9.2.x and v9.3.x before merging https://github.com/grafana/grafana/pull/58132 which is targeted for 9.4.x and changes a lot of things :smile: Given @kylebrandt is OOO until next week I might have to try picking this up so that I can get my other PRs merged, https://github.com/grafana/grafana/pull/58132, https://github.com/grafana/grafana/pull/58646

**Comment by kostrse** [comment]
@marefr OK, now it doesn't even attempt to call QueryData :)

```
Object
status:500
statusText:"Internal Server Error"
data:Object
error:"assignment to entry in nil map - C:/Users/user1/scoop/apps/go/current/src/runtime/map_faststr.go:205 (0x140015999)
	mapassign_faststr: panic(plainError(\"assignment to entry in nil map\"))
D:/Dev/grafana/pkg/expr/service/nodes.go:235 (0x141f60ba4)
	(*DSNode).Execute: req.Headers[\"Authorization\"] = fmt.Sprintf(\"%s %s\", token.Type(), token.AccessToken)
D:/Dev/grafana/pkg/expr/service/graph.go:54 (0x141f5e4c9)
	(*DataPipeline).execute: res, err := node.Execute(c, now, vars, s)
D:/Dev/grafana/pkg/expr/service/service.go:78 (0x141f637e8)
	(*Service).ExecutePipeline: vars, err := intPipeline.execute(ctx, now, s)
D:/Dev/grafana/pkg/expr/service/service.go:118 (0x141f63b87)
	(*Service).TransformData: responses, err := s.ExecutePipeline(ctx, now, pipeline)
D:/Dev/grafana/pkg/services/query/query.go:167 (0x1419cd5b1)
	(*Service).handleExpressions: qdr, err := s.expressionService.TransformData(ctx, time.Now(), &exprReq) // use time now because all queries have absolute time range
D:/Dev/grafana/pkg/services/query/query.go:89 (0x1419cccb1)
	(*Service).QueryData: return s.handleExpressions(ctx, user, parsedReq)
D:/Dev/grafana/pkg/api/metrics.go:57 (0x141dd358e)
	(*HTTPServer).QueryMetricsV2: resp, err := hs.queryDataService.QueryData(c.Req.Context(), c.SignedInUser, c.SkipCache, reqDTO)
D:/Dev/grafana/pkg/api/routing/routing.go:17 (0x14138a081)
	Wrap.func1: if res := handler(c); res != nil {
D:/Dev/grafana/pkg/api/response/web_hack.go:40 (0x140ddb735)
	wrap_handler.func3: handle(reqCtx(w, r))
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:135 (0x140d8d134)
	mwFromHandler.func1.1: h.ServeHTTP(mrw, r)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/services/contexthandler/contexthandler.go:186 (0x141372681)
	(*ContextHandler).Middleware.func1: next.ServeHTTP(w, r)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/render.go:44 (0x140d8dcf5)
	Renderer.func1.1: next.ServeHTTP(rw, req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/macaron.go:141 (0x140d8d1db)
	mwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/middleware/csrf/csrf.go:64 (0x141ad0d45)
	(*CSRF).Middleware.func1.1: next.ServeHTTP(w, r)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/middleware/recovery.go:171 (0x141384a21)
	Recovery.func1.1: next.ServeHTTP(w, req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/middleware/logger.go:40 (0x14138241c)
	Logger.func1.1: next.ServeHTTP(rw, r)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/middleware/request_metrics.go:58 (0x141385829)
	RequestMetrics.func1.1: next.ServeHTTP(w, r)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/middleware/request_tracing.go:88 (0x141386f7a)
	RequestTracing.func1.1: next.ServeHTTP(w, req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2109 (0x140341b0e)
	HandlerFunc.ServeHTTP: f(w, r)
D:/Dev/grafana/pkg/web/context.go:49 (0x140d8babd)
	(*Context).run: h.ServeHTTP(ctx.Resp, ctx.Req)
D:/Dev/grafana/pkg/web/router.go:155 (0x140d8f26d)
	(*Router).Handle.func1: c.run()
D:/Dev/grafana/pkg/web/router.go:211 (0x140d8fed5)
	(*Router).ServeHTTP: leaf.handle(rw, req, nil)
D:/Dev/grafana/pkg/web/macaron.go:167 (0x140d8d650)
	(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:2947 (0x140344f0b)
	serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
C:/Users/user1/scoop/apps/go/current/src/net/http/server.go:1991 (0x1403405e6)
	(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
C:/Users/user1/scoop/apps/go/current/src/runtime/asm_amd64.s:1594 (0x140071340)
	goexit: BYTE	$0x90	// NOP
"
message:"Internal Server Error - Check the Grafana server logs for the detailed error message."
config:Object
url:"api/ds/query?expression=true"
method:"POST"
data:Object
requestId:"Q118"
hideFromInspector:false
headers:Object
retry:0
message:"Internal Server Error - Check the Grafana server logs for the detailed error message."
```

**Comment by kostrse** [comment]
Repro:
1. Add `"oauthPassThru": true` to JsonData (any datasource, e.g. Prometheus)
2. Add math expression to panel
3. QueryData won't be executed and Internal Server Error will be shown.

**Comment by papagian** [comment]
> @marefr @papagian looks like this is good to go, but is blocked? Would it also be possible to port this to 9.1, 9.2 (with a slight preference to 9.2)?

I don't think it's blocked by me; I'm assigned as a reviewer on behalf of the backend platform
but I'm afraid that I'm not a lot in sync with the latest changes regarding expressions & query-ring datasources .
However, if you still think that my review would be beneficial let me know.

**Comment by kostrse** [comment]
@marefr Now it seems to be working fine.
Tested with several internal and external datasources.

**Review by wbrowne — APPROVED** [review]
LGTM

**Pull request merged**